Source/Core/InputCommon/ImageOperations.cpp

```diff
@@ -0,0 +1,381 @@
+// Copyright 2019 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "ImageOperations.h"
+
+#include <algorithm>
+#include <cmath>
+#include <limits>
+#include <stack>
+
+#include <png.h>
+
+#include "Common/File.h"
+#include "Common/FileUtil.h"
+#include "Common/Image.h"
+
+namespace InputCommon
+{
+namespace
+{
+Pixel SampleNearest(const ImagePixelData& src, double u, double v)
+{
+  const u32 x = std::clamp(static_cast<u32>(u * src.width), 0u, src.width - 1);
+  const u32 y = std::clamp(static_cast<u32>(v * src.height), 0u, src.height - 1);
+  return src.pixels[x + y * src.width];
+}
+
+void CalculateNeighboringPixels(const ImagePixelData& src, std::vector<Point>& neighboring_pixels,
+                                std::vector<bool>& seen_pixels, u32 x, u32 y,
+                                const Pixel& matching_color)
+{
+  std::stack<Point> pixels_to_test;
+  auto add_pixel_if_valid = [&](u32 x, u32 y) {
+    if (x < 0)
+    {
+      return;
+    }
+
+    if (y < 0)
+    {
+      return;
+    }
+
+    if (x >= src.width)
+    {
+      return;
+    }
+
+    if (y >= src.height)
+    {
+      return;
+    }
+
+    if (seen_pixels[y * src.width + x])
+    {
+      return;
+    }
+
+    const auto pixel = src.pixels[y * src.width + x];
+    if (pixel == matching_color)
+    {
+      pixels_to_test.emplace(Point{x, y});
+      seen_pixels[y * src.width + x] = true;
+      neighboring_pixels.emplace_back(Point{x, y});
+    }
+  };
+
+  pixels_to_test.emplace(Point{x, y});
+  while (!pixels_to_test.empty())
+  {
+    const auto pixel = pixels_to_test.top();
+    pixels_to_test.pop();
+
+    add_pixel_if_valid(pixel.x - 1, pixel.y);
+    add_pixel_if_valid(pixel.x - 1, pixel.y + 1);
+    add_pixel_if_valid(pixel.x - 1, pixel.y - 1);
+    add_pixel_if_valid(pixel.x + 1, pixel.y);
+    add_pixel_if_valid(pixel.x + 1, pixel.y + 1);
+    add_pixel_if_valid(pixel.x + 1, pixel.y - 1);
+    add_pixel_if_valid(pixel.x, pixel.y - 1);
+    add_pixel_if_valid(pixel.x, pixel.y + 1);
+  }
+}
+}  // namespace
+
+void CopyTo(const ImagePixelData& src, ImagePixelData& dst, const Rect& src_region,
+            const Rect& dst_region)
+{
+  if (src_region.width != dst_region.width || src_region.height != dst_region.height)
+  {
+    return;
+  }
+
+  for (u32 x = 0; x < dst_region.width; x++)
+  {
+    for (u32 y = 0; y < dst_region.height; y++)
+    {
+      dst.pixels[(y + dst_region.p.y) * dst.width + x + dst_region.p.x] =
+          src.pixels[(y + src_region.p.y) * src.width + x + src_region.p.x];
+    }
+  }
+}
+
+std::optional<ImagePixelData> LoadImage(const std::string& path)
+{
+  File::IOFile file;
+  file.Open(path, "rb");
+  std::vector<u8> buffer(file.GetSize());
+  file.ReadBytes(buffer.data(), file.GetSize());
+
+  ImagePixelData image;
+  std::vector<u8> data;
+  if (!Common::LoadPNG(buffer, &data, &image.width, &image.height))
+    return {};
```

Comment:
`return std::nullopt;`

The nullopt constructor and default constructor invocations don't always do the [same thing](https://godbolt.org/z/UHP5B9)

Reply:
Very interesting, I wasn't aware of that.  Thank you!